File: easy/Two Sum/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  const hashMap = {};
  const n = nums.length;

  for (let i = 0; i < n; ++i) {
    const toFind = target - nums[i];
    if (hashMap[toFind] != undefined) {
      return [i, hashMap[toFind]];
    } else {
      hashMap[nums[i]] = i;
    }
  }
};

const testSum = function (nums, target, output) {
  const result = twoSum(nums, target);
  if (result == output) {
    console.log("Test Case Passed!");
  } else {
    console.log(
      `Test Case Failed: nums: ${nums} target: ${target} output: ${output} result: ${result}`
    );
  }
};

testSum([2, 7, 11, 15], 9, [1, 0]);
testSum([3, 2, 4], 6, [2, 1]);
testSum([3, 3], 6, [1, 0]);

File: second_pass/1_arrays_and_hashing/3_two_sum/two_sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = {}
        for i, num in enumerate(nums):
            map[num] = [i]

        for i, num in enumerate(nums):
            to_find = target - num
            if to_find in map:
                index = map[to_find]
                return [index, i]

        return None

    def testTwoSum(self, nums: List[int], target: int, expected: int):
        actual = self.twoSum(nums, target)
        if actual == expected:
            print("Test Case Passed!")
        else:
            print(f"Test Case Failed. nums: {nums} target: {target} actual: {actual} expected:{expected}")
